lib/base/json.cpp

```diff
@@ -2,22 +2,253 @@
 
 #include "base/json.hpp"
 #include "base/debug.hpp"
-#include "base/namespace.hpp"
-#include "base/dictionary.hpp"
 #include "base/array.hpp"
-#include "base/objectlock.hpp"
-#include "base/convert.hpp"
+#include "base/generator.hpp"
 #include "base/utility.hpp"
-#include <bitset>
-#include <boost/exception_ptr.hpp>
-#include <cstdint>
-#include <json.hpp>
 #include <stack>
 #include <utility>
 #include <vector>
 
 using namespace icinga;
 
+constexpr uint8_t JsonEncoder::m_IndentSize;
+
+JsonEncoder::JsonEncoder(std::string& output, bool prettify)
+	: JsonEncoder{nlohmann::detail::output_adapter<char>(output), prettify}
+{
+}
+
+JsonEncoder::JsonEncoder(std::basic_ostream<char>& stream, bool prettify)
+	: JsonEncoder{nlohmann::detail::output_adapter<char>(stream), prettify}
+{
+}
+
+JsonEncoder::JsonEncoder(nlohmann::detail::output_adapter_t<char> stream, bool prettify)
+	: m_Pretty(prettify), m_Indent{0}, m_IndentStr(32, ' '), m_Writer(std::move(stream))
+{
+}
+
+/**
+ * Encodes the given value into JSON and writes it to the configured output stream.
+ *
+ * This function is specialized for the @c Value type and recursively encodes each
+ * and every concrete type it represents.
+ *
+ * @param value The value to be JSON serialized.
+ */
+void JsonEncoder::EncodeImpl(const Value& value, std::optional<boost::asio::yield_context>& yc, bool parentOlockAcquired)
+{
+	switch (value.GetType()) {
+		case ValueEmpty:
+			Write("null");
+			return;
+		case ValueBoolean:
+			Write(value.ToBool() ? "true" : "false");
+			return;
+		case ValueString:
+			EncodeStringOrNumber(value.Get<String>());
+			return;
+		case ValueNumber:
+			if (auto ll(static_cast<long long>(value)); ll == value) {
+				EncodeStringOrNumber(ll);
+			} else {
+				EncodeStringOrNumber(value.Get<double>());
+			}
+			return;
+		case ValueObject: {
+			const Object::Ptr& obj = value.Get<Object::Ptr>();
+			const auto& type = obj->GetReflectionType();
```

Comment:
I don't really understand the pattern in which `auto` is used here. I'd probably use it in both cases, especially as the first line contains the type twice at the moment.